File: mitopulse_mobile_v1/mobile/ios/MitoPulseIOS/Sources/Engine.swift
import Foundation
import CryptoKit

struct Sample {
  let ts: Int
  let hr: Double
  let hrvRmssd: Double?
  let spo2: Double?
  let sleepScore: Double?
  let accelLoad: Double?
  let altitudeM: Double?
  let tempC: Double?
  let humidity: Double?
}

struct ComputeResult {
  let indexValue: Double
  let slope: Double
  let tier: String
  let risk: Int
  let coercion: Bool
  let dynamicId: String
  let vectorJson: String
}

enum Engine {
  static func clamp(_ x: Double,_ lo: Double,_ hi: Double)->Double { max(lo, min(hi, x)) }
  static func tier(_ s: Sample)->String {
    let hasBasic = (s.spo2 != nil || s.sleepScore != nil || s.accelLoad != nil)
    return (hasBasic && s.hrvRmssd != nil) ? "tier2" : "tier1"
  }
  static func cEnv(_ s: Sample)->Double {
    if s.altitudeM==nil && s.tempC==nil && s.humidity==nil { return 1.0 }
    let alt=(s.altitudeM ?? 0.0)/1000.0
    let t=s.tempC ?? 22.0
    let h=s.humidity ?? 50.0
    let v=1.0/(1.0 + 0.012*alt + 0.008*abs(t-22.0) + 0.005*abs(h-50.0)/50.0)
    return clamp(v, 0.85, 1.15)
  }
  static func norm(_ s: Sample)->[String:Double] {
    var out:[String:Double]=[:]
    out["hr"]=clamp(1.0-(s.hr-50.0)/80.0,0.0,1.0)
    if let h=s.hrvRmssd { out["hrv"]=clamp((h-10.0)/90.0,0.0,1.0) }
    if let o=s.spo2 { out["spo2"]=clamp((o-90.0)/10.0,0.0,1.0) }
    if let sl=s.sleepScore { out["sleep"]=clamp(sl/100.0,0.0,1.0) }
    if let ld=s.accelLoad { out["load"]=clamp(1.0-ld/10.0,0.0,1.0) }
    return out
  }
  static func hmac(secret: Data, msg: Data)->String {
    let key = SymmetricKey(data: secret)
    let mac = HMAC<SHA256>.authenticationCode(for: msg, using: key)
    return Data(mac).base64EncodedString().replacingOccurrences(of: "+", with: "-").replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "")
  }
  static func compute(history: inout [Double], s: Sample, secret: Data)->ComputeResult {
    let t = tier(s)
    let b = norm(s)
    let w:[String:Double] = (t=="tier2") ? ["hr":0.30,"hrv":0.30,"spo2":0.20,"sleep":0.10,"load":0.10] : ["hr":0.40,"spo2":0.25,"sleep":0.20,"load":0.15]
    var acc=0.0; var tw=0.0
    for (k,wk) in w { if let v=b[k] { acc += wk*v; tw += wk } }
    let base = (tw<=0) ? 0.5 : acc/tw
    let idx = clamp(base*cEnv(s),0.0,1.0)
    history.append(idx)
    let mean = history.reduce(0,+)/Double(history.count)
    let last = history.last ?? idx
    let slope = (history.count>=2) ? (history[history.count-1]-history[history.count-2]) : 0.0
    let vec:[String:Double] = ["mean":mean,"last":last,"slope":slope]
    let vecData = try! JSONSerialization.data(withJSONObject: vec, options: [.sortedKeys])
    let dyn = hmac(secret: secret, msg: vecData)
    var risk=0
    if idx<0.25 { risk += 50 }
    if slope < -0.05 { risk += 20 }
    if let h=s.hrvRmssd, h<20 { risk += 20 }
    if s.hr>110 { risk += 20 }
    risk = Int(clamp(Double(risk),0,100))
    let coercion = risk >= 75
    let vecJson = String(data: vecData, encoding: .utf8) ?? "{}"
    return ComputeResult(indexValue: idx, slope: slope, tier: t, risk: risk, coercion: coercion, dynamicId: dyn, vectorJson: vecJson)
  }
}

File: mitopulse_v3/MitoPulse_v3_GLOBAL_DEMO/mobile/android/MitoPulseAndroid/app/src/main/java/com/gmative/mitopulse/Engine.kt
package com.gmative.mitopulse
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

data class Sample(
  val ts: Long,
  val hr: Double,
  val hrvRmssd: Double?=null,
  val spo2: Double?=null,
  val sleepScore: Double?=null,
  val accelLoad: Double?=null,
  val altitudeM: Double?=null,
  val tempC: Double?=null,
  val humidity: Double?=null,
)

data class ComputeResult(
  val indexValue: Double,
  val slope: Double,
  val tier: String,
  val risk: Int,
  val coercion: Boolean,
  val dynamicId: String,
  val vectorJson: String
)

object Engine {
  private fun clamp(x: Double, lo: Double, hi: Double)=max(lo, min(hi, x))
  private fun tier(s: Sample): String {
    val hasBasic = (s.spo2!=null || s.sleepScore!=null || s.accelLoad!=null)
    return if (hasBasic && s.hrvRmssd!=null) "tier2" else "tier1"
  }
  private fun cEnv(s: Sample): Double {
    if (s.altitudeM==null && s.tempC==null && s.humidity==null) return 1.0
    val alt=(s.altitudeM?:0.0)/1000.0
    val t=s.tempC?:22.0
    val h=s.humidity?:50.0
    val v=1.0/(1.0 + 0.012*alt + 0.008*abs(t-22.0) + 0.005*abs(h-50.0)/50.0)
    return clamp(v, 0.85, 1.15)
  }
  private fun norm(s: Sample): Map<String, Double> {
    val out=mutableMapOf<String,Double>()
    out["hr"]=clamp(1.0-(s.hr-50.0)/80.0,0.0,1.0)
    s.hrvRmssd?.let { out["hrv"]=clamp((it-10.0)/90.0,0.0,1.0) }
    s.spo2?.let { out["spo2"]=clamp((it-90.0)/10.0,0.0,1.0) }
    s.sleepScore?.let { out["sleep"]=clamp(it/100.0,0.0,1.0) }
    s.accelLoad?.let { out["load"]=clamp(1.0-it/10.0,0.0,1.0) }
    return out
  }
  fun compute(history: MutableList<Double>, s: Sample, secret: ByteArray): ComputeResult {
    val tierUsed=tier(s)
    val b=norm(s)
    val w = if (tierUsed=="tier2") mapOf("hr" to 0.30,"hrv" to 0.30,"spo2" to 0.20,"sleep" to 0.10,"load" to 0.10)
            else mapOf("hr" to 0.40,"spo2" to 0.25,"sleep" to 0.20,"load" to 0.15)
    var acc=0.0; var tw=0.0
    for ((k,wk) in w) {
      val v=b[k] ?: continue
      acc += wk*v; tw += wk
    }
    val base = if (tw<=0.0) 0.5 else acc/tw
    val idx = clamp(base*cEnv(s), 0.0, 1.0)
    history.add(idx)
    val mean=history.sum()/history.size
    val last=history.last()
    val slope= if (history.size>=2) history[history.size-1]-history[history.size-2] else 0.0
    val vecJson = "{"mean":$mean,"last":$last,"slope":$slope}"
    val dyn = Crypto.hmacSha256(secret, vecJson.toByteArray(Charsets.UTF_8))
    var risk=0
    if (idx<0.25) risk+=50
    if (slope<-0.05) risk+=20
    if (s.hrvRmssd!=null && s.hrvRmssd<20) risk+=20
    if (s.hr>110) risk+=20
    risk=clamp(risk.toDouble(),0.0,100.0).toInt()
    val coercion=risk>=75
    return ComputeResult(idx,slope,tierUsed,risk,coercion,dyn,vecJson)
  }
}
